eclipse/che
# Run dashboard in a dedicated deployment with helm chart

### What does this PR do?
Run dashboard in a dedicated deployment.
Note that after this PR Che Server tomcat will continue serving dashboard resources as well due backward compatibility reason. Tomcat will stop serving dashboard resources only after two sprints after we move helm  + che operator to new deployment way.

### Screenshot/screencast of this PR
![Screenshot_20201012_154933](https://user-images.githubusercontent.com/5887312/95748376-8f19cc80-0ca2-11eb-911b-f09d674f26cb.png)

### What issues does this PR fix or reference?
It fixes https://github.com/eclipse/che/issues/18076

### How to test this PR?
I checked that Dashboard and Dashboard Next are accessible and work as expected in different combination Singlehost/Multihost/DefaultHost/Gateway + Single/Multi-user

* note that I generated certificates with chectl before I executed helm command: (`chectl server:start --platform=minikube --installer=helm && helm delete che`)
* custom che server image is used to make sure that dashboard is loaded from Apache server but not Tomcat;

<details>
<summary>Singlehost</summary>

<details>
<summary>Singlehost Single-User Che</summary>

```shell
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
  --force \
  --namespace che \
  --set global.cheDomain=192.168.99.100.nip.io \
  --set global.ingressDomain=192.168.99.100.nip.io \
  --set cheImage=sleshchenko/che-server:cut-dashboard \
  --set global.serverStrategy=single-host \
  --set global.tls.useSelfSignedCerts=true \
  -f ${CHART_DIR}/values/tls.yaml \
  ${CHART_DIR}
```
</details>

<details>
<summary>Singlehost Mutli-User Che</summary>

```shell
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
  --force \
  --namespace che \
  --set global.cheDomain=192.168.99.100.nip.io \
  --set global.ingressDomain=192.168.99.100.nip.io \
  --set cheImage=sleshchenko/che-server:cut-dashboard \
  --set global.serverStrategy=single-host \
  --set global.tls.useSelfSignedCerts=true \
  -f ${CHART_DIR}/values/tls.yaml \
  -f ${CHART_DIR}/values/multi-user.yaml \
  ${CHART_DIR}
```

</details>

</details>

<details>
  <summary>SingleHost Gateway</summary>

  <details>
  <summary>SingleHost Gateway Single-User Che</summary>
  
```shell
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
    --force \
    --namespace che \
    --set global.cheDomain=192.168.99.100.nip.io \
    --set global.ingressDomain=192.168.99.100.nip.io \
    --set cheImage=sleshchenko/che-server:cut-dashboard \
    --set global.tls.useSelfSignedCerts=true \
    --set global.serverStrategy=single-host \
    --set cheSinglehostGateway.deploy=true \
    --set global.singleHostExposure=gateway \
    -f ${CHART_DIR}/values/tls.yaml \
    ${CHART_DIR}
```
  </details>

  <details>
  <summary>SingleHost Gateway Mutli-User Che</summary>

```shell
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
    --force \
    --namespace che \
    --set global.cheDomain=192.168.99.100.nip.io \
    --set global.ingressDomain=192.168.99.100.nip.io \
    --set cheImage=sleshchenko/che-server:cut-dashboard \
    --set global.tls.useSelfSignedCerts=true \
    --set global.serverStrategy=single-host \
    --set cheSinglehostGateway.deploy=true \
    --set global.singleHostExposure=gateway \
    -f ${CHART_DIR}/values/tls.yaml \
    -f ${CHART_DIR}/values/multi-user.yaml \
    ${CHART_DIR}
```

</details>

</details>



<details>
  <summary>Multi Host</summary>

  <details>
  <summary>MultiHost Single-User Che</summary>

```shell
  export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
  helm upgrade --install che \
  --force \
  --namespace che \
  --set global.cheDomain=192.168.99.100.nip.io \
  --set global.ingressDomain=192.168.99.100.nip.io \
  --set cheImage=sleshchenko/che-server:cut-dashboard \
  --set global.serverStrategy=multi-host \
  --set global.tls.useSelfSignedCerts=true \
  -f ${CHART_DIR}/values/tls.yaml \
  ${CHART_DIR}
```
  </details>

  <details>
  <summary>MultiHost Mutli-User Che</summary>

```shell
  export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
  helm upgrade --install che \
    --force \
    --namespace che \
    --set global.cheDomain=192.168.99.100.nip.io \
    --set global.ingressDomain=192.168.99.100.nip.io \
    --set cheImage=sleshchenko/che-server:cut-dashboard \
    --set global.serverStrategy=multi-host \
    --set global.tls.useSelfSignedCerts=true \
    -f ${CHART_DIR}/values/tls.yaml \
    -f ${CHART_DIR}/values/multi-user.yaml \
    ${CHART_DIR}
```

  </details>

  </details>

<details>
  <summary>Default Host</summary>

Note: it's without TLS, seems that does not work with TLS properly.

  <details>
  <summary>Default Host Single-User Che</summary>

```sh
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
    --force \
    --namespace che \
    --set global.cheDomain=192.168.99.100 \
    --set global.ingressDomain=192.168.99.100 \
    --set cheImage=sleshchenko/che-server:cut-dashboard \
    --set global.tls.useSelfSignedCerts=false \
    --set global.serverStrategy=default-host \
${CHART_DIR}
```
  </details>

  <details>
  <summary>Default Host Mutli-User Che</summary>

```shell
export CHART_DIR=/home/sleshche/projects/che/deploy/kubernetes/helm/che
helm upgrade --install che \
    --force \
    --namespace che \
    --set global.cheDomain=192.168.99.100 \
    --set global.ingressDomain=192.168.99.100 \
    --set cheImage=sleshchenko/che-server:cut-dashboard \
    --set global.tls.useSelfSignedCerts=false \
    --set global.serverStrategy=default-host \
    -f ${CHART_DIR}/values/multi-user.yaml \
    ${CHART_DIR}
```

  </details>

  </details>


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
